eventmesh-common/src/main/java/org/apache/eventmesh/common/protocol/tcp/codec/Codec.java

```diff
@@ -48,139 +46,122 @@
-
-        @Override
-        public void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) throws Exception {
-            try {
-                if (null == in) {
-                    return;
-                }
-
-                byte[] flagBytes = parseFlag(in);
-                byte[] versionBytes = parseVersion(in);
-                validateFlag(flagBytes, versionBytes, ctx);
-
-                final int length = in.readInt();
-                final int headerLength = in.readInt();
-                final int bodyLength = length - CONSTANT_MAGIC_FLAG.length - VERSION.length - headerLength;
-                Header header = parseHeader(in, headerLength);
-                Object body = parseBody(in, header, bodyLength);
-
-                Package pkg = new Package(header, body);
-                out.add(pkg);
-            } catch (Exception e) {
-                log.error("decode error| received data: {}.", deserializeBytes(in.array()), e);
-                throw e;
-            }
-        }
+        int headerLength = ArrayUtils.getLength(headerData);
+        int bodyLength = ArrayUtils.getLength(bodyData);
+
+        final int length = CONSTANT_MAGIC_FLAG.length + VERSION.length + headerLength + bodyLength;
 
-        private byte[] parseFlag(ByteBuf in) {
-            final byte[] flagBytes = new byte[CONSTANT_MAGIC_FLAG.length];
-            in.readBytes(flagBytes);
-            return flagBytes;
+        if (length > FRAME_MAX_LENGTH) {
+            throw new IllegalArgumentException("message size is exceed limit!");
+        }
+        /**
+         * Header + Body, Format:
+         * <pre>
+         * ┌───────────────┬─────────────┬──────────────────┬──────────────────┬──────────────────┬─────────────────┐
+         * │   MAGIC_FLAG  │   VERSION   │ package length   │   Header length  │      Header      │      body       │
+         * │    (9bytes)   │  (4bytes)   │    (4bytes)      │      (4bytes)    │   (header bytes) │   (body bytes)  │
+         * └───────────────┴─────────────┴──────────────────┴──────────────────┴──────────────────┴─────────────────┘
+         * </pre>
+         */
+        out.writeBytes(CONSTANT_MAGIC_FLAG);
+        out.writeBytes(VERSION);
+        out.writeInt(length);
+        out.writeInt(headerLength);
+        if (headerData != null) {
+            out.writeBytes(headerData);
         }
+        if (bodyData != null) {
+            out.writeBytes(bodyData);
+        }
+    }
 
-        private byte[] parseVersion(ByteBuf in) {
-            final byte[] versionBytes = new byte[VERSION.length];
-            in.readBytes(versionBytes);
-            return versionBytes;
+    @Override
+    protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) throws Exception {
+        if (in == null) {
+            return;
+        }
+        if (in.readableBytes() < CONSTANT_MAGIC_FLAG.length + VERSION.length + 4 + 4) {
```

Comment:
> Could these magic numbers be changed to constants to indicate that they are the length of the package length value and the length of the header length value?

These magic numbers are actually part of the protocol. I recommend keeping them. You can think of them as a special marker for EventMesh's custom protocol, similar to the magic number in Java class files.

Reply:
> Could these 2 magic numbers be changed to constants to indicate that they are the length of the package length value and the length of the header length value?

Using constants can improve readability and maintainability. But in this case, the numbers are integral to the protocol's structure, not arbitrary choices. It is better to stick with protocol-defined markers, otherwise it could lead to ambiguity or unexpected behavior.